### ISSUE TITLE
hide commands dynamically

### ISSUE THREAD

**Issue by eqvis** [remkop/picocli#1478]
Hi,
is there a way to hide commands dynamically?
I use picocli with jline3 and want to show some commands only if a connection with a server is established.

Best regards,
Mayra

**Comment by remkop** [comment]
Hi @eqvis,

One idea is to use a custom `IHelpSectionRenderer`, see this example: https://github.com/remkop/picocli/blob/master/picocli-examples/src/main/java/picocli/examples/customhelp/ConditionallyShowSubcommands.java

Another idea is to use a custom `IHelpFactory`, see this example: https://github.com/remkop/picocli/blob/master/picocli-examples/src/main/java/picocli/examples/customhelp/AlphabeticSubcommands.java

I hope this helps. 😅

**Comment by eqvis** [comment]
Ah, yeah, thank you! To remove them from TAB-completion I would have to extend PicocliJLineCompleter, right?

**Comment by remkop** [comment]
Yes, that is one idea. Partial completion candidates may be a bit harder to filter out, but probably doable. 

For example, you want to filter out subcommand "mysub". Assume that the user already typed "my", then hits `<TAB>`. The completion candidates will now include "sub" (the remainder of the "mysub" command).
So you may need to prefix the returned completion candidates with the partial user input so far, to get the full candidates, then filter out the commands you want to hide, and finally return the remaining completion candidates.

**Comment by remkop** [comment]
Hi @eqvis, do you have any follow-up questions on this, or can this ticket be closed?

**Comment by eqvis** [comment]
Hi @remkop, well it would be cool if there was a way to do this out of the box (like passing a callable to the hidden parameter…), but I guess it's only useful when using picocli with jline. At the moment, we go with the solution of showing a message in the status-pane.
Thank you for your support,
Mayra

**Comment by remkop** [comment]
@eqvis Just took another look at this issue again (thanks for your patience), and had this idea:
have you tried changing the `hidden` attribute on the usage help message?
Internally picocli looks at that attribute for all kinds of things including autocompletion.

```java
CommandLine commandLine = myCommandLine();
CommandSpec spec = commandLine.getCommandSpec();
spec.usageMessage().hidden(true);
```
